examples/generative/Density estimation using RealNVP

```diff
@@ -0,0 +1,245 @@
+        self.layer10 = Dense(input_shape, activation="tanh", kernel_regularizer=l2(reg))
+
+    # Operations performed when the layer is called
+
+    def call(self, x):
+
+        # Calling the previously specified layers one by one
+        y = self.layer1(x)
+        y = self.layer2(y)
+        y = self.layer3(y)
+        y = self.layer4(y)
+        t = self.layer5(y)
+
+        z = self.layer6(x)
+        z = self.layer7(z)
+        z = self.layer8(z)
+        z = self.layer9(z)
+        s = self.layer10(z)
+
+        return s, t
+
+
+"""
+## Real NVP
+"""
+
+# class containing the whole set of operations
+
+
+class realnvp(Layer):
+    def __init__(self, layers, masks, distr):
+        super(realnvp, self).__init__()
+
+        # number of coupling layers
+        self.num_cl = len(masks)
+        # distribution of the latent space
+        self.distr = distr
+        # masks to divide into 1:d and d+1:D
+        self.masks = masks
+        # s and t custom layers of before
+        self.layers = layers
+
+    # Custom function defining the forward operation
+
+    def forward(self, y):
+        # log determinant of the forward pass
+        log_det_for = tf.zeros(y.shape[0])
+        x = y
+        for i in range(self.num_cl):
+            x_masked = x * self.masks[i]
+            reversed_mask = 1 - self.masks[i]
+            s, t = self.layers[i](x_masked)
+            x = x_masked + reversed_mask * (
+                x * tf.exp(s) * reversed_mask + t * reversed_mask
+            )
+            log_det_for += tf.reduce_sum(s, [1])
+
+        return x, log_det_for
+
+    # Custom function defining the forward operation
+
+    def inverse(self, x):
+        # log determinant of the forward pass
+        log_det_inv = tf.zeros(x.shape[0])
+        y = x
+        for i in reversed(range(self.num_cl)):
+            y_masked = y * self.masks[i]
+            reversed_mask = 1 - self.masks[i]
+            s, t = self.layers[i](y_masked)
+            y = (
+                reversed_mask * (y - t * reversed_mask) * tf.exp(-(s * reversed_mask))
+                + y_masked
+            )
+            log_det_inv -= tf.reduce_sum(s, [1])
+
+        return y, log_det_inv
+
+    # log likelihood of the normal distribution + the log determinant of the jacobian
+
+    def log_likelihood(self, x):
+        y, logdet = self.inverse(x)
+        return self.distr.log_prob(y) + logdet
+
+
+"""
+## Initialization
+"""
+
+# initializing the masks, the layers and the prior distribution for the real nvp
+num_cl = 6
+masks = np.array([[0, 1], [1, 0]] * (num_cl // 2), dtype="float32")
+layers = [s_t(2) for i in range(num_cl)]
+
+tfd = tfp.distributions
+distr = tfd.MultivariateNormalDiag(loc=[0.0, 0.0], scale_diag=[1.0, 1.0])
+
+real_nvp = realnvp(layers, masks, distr)
+
+"""
+## Model training
+"""
+
+loss_history = []
+# Adam optimizer
+optimizer = tf.keras.optimizers.Adam(learning_rate=0.00005)
+
+# 2000 epochs
+for i in notebook.tqdm(range(2000)):
```

Comment:
Let's turn this logic into a more idiomatic Keras workflow -- please factor this into a `train_step` method on a Model subclass -- see https://keras.io/guides/customizing_what_happens_in_fit/